pkgs/build-support/node/build-npm-package/default.nix

```diff
@@ -0,0 +1,52 @@
+{ lib, stdenv, fetchNpmDeps, npmHooks, nodejs }:
+
+{ name ? "${args.pname}-${args.version}"
+, src ? null
+, srcs ? null
+, sourceRoot ? null
+, patches ? [ ]
+, nativeBuildInputs ? [ ]
+, buildInputs ? [ ]
+  # The output hash of the dependencies for this project.
+  # Can be calculated in advance with prefetch-npm-deps.
+, npmDepsHash ? ""
+  # Whether to make the cache writable prior to installing dependencies.
+  # Don't set this unless npm tries to write to the cache directory, as it can slow down the build.
+, makeCacheWritable ? false
+  # The script to run to build the project.
+, npmBuildScript ? "build"
+  # Flags to pass to all npm commands.
+, npmFlagsArray ? [ ]
+  # Flags to pass to `npm ci`.
+, npmInstallFlagsArray ? [ ]
+  # Flags to pass to `npm run ${npmBuildScript}`.
+, npmBuildFlagsArray ? [ ]
+  # Flags to pass to `npm pack`.
+, npmPackFlagsArray ? [ ]
+, ...
+} @ args:
+
+let
+  npmDeps = fetchNpmDeps {
+    inherit src srcs sourceRoot patches;
+    name = "${name}-npm-deps";
+    hash = npmDepsHash;
+  };
+
+  inherit (npmHooks.override { inherit nodejs; }) npmConfigHook npmBuildHook npmInstallHook;
+in
+stdenv.mkDerivation (args // {
```

Comment:
See https://github.com/NixOS/nixpkgs/pull/189539#discussion_r990434397, all these list arguments should be escaped with `lib.escapeShellArgs` before being propagated to the `mkDerivation` call.

Reply:
My only issue with this approach is that it makes the UX poor for people who need to set flags while using the individual hooks -- how will they know to escape the flags, other than by documenting it?

Reply:
Not sure what you mean, got an example?

Reply:
Sure:
```nix
buildGoModule rec {
  pname = "evcc";
  version = "0.104.2";

  src = fetchFromGitHub {
    owner = "evcc-io";
    repo = pname;
    rev = version;
    hash = "sha256-iTU0E8cWDUmTUxWMXW6zsbA0p+F9oyF3RLei73A6BrI=";
  };

  vendorHash = "sha256-tkGKKtPUTGlUpq5V9QzqON9LotAlUeeV2EwmdZ++VNk=";

  npmDeps = fetchNpmDeps {
    inherit src;
    hash = "sha256-qBzj+SfAf41Ew5jUBRrNWVIwTkMpl39UOJ20zjVGaDg=";
  };

  nativeBuildInputs = [ nodejs npmHooks.npmConfigHook ];

  npmInstallFlags = [ "--legacy-peer-deps" ];

  preBuild = ''
    make ui
  '';

  doCheck = !stdenv.isDarwin; # tries to bind to local network, doesn't work in darwin sandbox
}
```

How would the user know to `lib.escapeShellArgs` the value of `npmInstallFlags`?

Reply:
Ahh I see, so in cases where the `buildNpmPackage` function isn't used, but only the `npmConfigHook`.

How about this: Change `nmpConfigHook` to take the `npmInstallFlags` argument instead, then it can properly handle escaping itself.

Reply:
So make `npmConfigHook` a function that returns a hook? That's... pretty non-standard, and deviates from how users expect to be able to add hooks/pass parameters to them.

It's a good idea, but I'm not sure if it's the right solution here.

Reply:
I feel like the best route here is to document that `lib.escapeStrings` should be used in this context, or breakages will happen. I think that's really all we can do, right?

Better than nothing, at least...

Reply:
I'm really not sure. I think the hook taking arguments is much clearer. However it is a new way of doing things, and in the [Nixpkgs Architecture Team](https://nixos.org/community/teams/nixpkgs-architecture.html) there have been discussions around improving the state-of-the art for this kind of thing, so it would be a shame for this somewhat new idea to be invalidated soonish. But then again, hooks taking arguments isn't a revolutionary new idea, it's just combining existing mechanisms in a new way. And btw, there's at least one hook and one usage of that hook which does take arguments: https://github.com/NixOS/nixpkgs/blob/678b994771c7fadfb3d7d3d96ee63fdf6fc574d0/pkgs/tools/filesystems/ceph/default.nix#L159